### PR TITLE
Infer entity id type

### PR DESCRIPTION
**Pull request by ezzabuzaid**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/elf/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[x] Other... Please describe: This PR is a completion of the work that @5im0n did
```

## What is the current behavior?

Attempt to add Entities Action in Elf

Issue Number: #247 

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/store/src/lib/store.ts [hunk 1]
```diff
@@ -1,21 +1,30 @@
-import { BehaviorSubject, Observable } from 'rxjs';
-import { Query } from '..';
-import { batchInProgress, batchDone$ } from './batch';
+import { BehaviorSubject, Observable, Subject } from 'rxjs';
+import { Action, Query } from '..';
+import { batchDone$, batchInProgress } from './batch';
 import { elfHooksRegistry } from './elf-hooks';
 import { addStore, removeStore } from './registry';
 
+type IdKeyType<SDef extends StoreDef<any>> =
```
Comment (NetanelBasal) [inline packages/store/src/lib/store.ts:7]: The store doesn't need to know anything about entities.

File: packages/store/src/lib/actions.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+export enum EntityActions {
```
Comment (NetanelBasal) [inline packages/store/src/lib/actions.ts:1]: The base store doesn't need to know anything about entities.

Comment (ezzabuzaid) [reply packages/store/src/lib/actions.ts:1]: I've changed the name to `EntityActions` to ease the migration from Akita. I guess that's better, but I can change it back. let me know...

Comment (NetanelBasal) [reply packages/store/src/lib/actions.ts:1]: EntityActions should be in the entities package

File: packages/store/src/lib/actions.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  Set = 'Set',
+  Add = 'Add',
+  Update = 'Update',
+  Remove = 'Remove',
+}
+
+export interface Action<IDType> {
```
Comment (NetanelBasal) [inline packages/store/src/lib/actions.ts:8]: The base action should have only a `type`.

Comment (NetanelBasal) [reply packages/store/src/lib/actions.ts:8]: We can add a payload key:

export interface Action<T extends string, Payload> {
  type: T,
  payload: Payload
}

Comment (ezzabuzaid) [reply packages/store/src/lib/actions.ts:8]: I'd still need to use the `Action` type in the store class, I'm not sure I can think of a way where store and entities are decoupled.

```ts
actions$: Observable<Action<EntityActions, EntityIdType>> =
    this.context.actions.asObservable();
```

The same problem with inferring the EntityIdType, the store has to have a bit of knowledge about the entities to figure out the correct type.

I'm not that great with typescript so any help is appreciated!

Comment (NetanelBasal) [reply packages/store/src/lib/actions.ts:8]: We can probably leave it as `Observable<Action<any, any>>` and let the `ofType` function to narrow down the type

**Pull request closed without merging**